diff -urN /home/zzoon.ko/opensrc/tarball/gst-rtsp-server-1.2.3/gst/rtsp-server/Makefile.am ./Makefile.am
--- /home/zzoon.ko/opensrc/tarball/gst-rtsp-server-1.2.3/gst/rtsp-server/Makefile.am	2014-01-13 01:50:54.000000000 +0900
+++ ./Makefile.am	2015-01-15 16:49:47.091243189 +0900
@@ -7,6 +7,7 @@
 		rtsp-thread-pool.h \
 		rtsp-media.h \
 		rtsp-media-factory.h \
+		rtsp-media-factory-wfd.h \
 		rtsp-media-factory-uri.h \
 		rtsp-mount-points.h \
 		rtsp-permissions.h \
@@ -16,8 +17,11 @@
 		rtsp-session-media.h \
 		rtsp-session-pool.h \
 		rtsp-token.h \
+		rtsp-client-wfd.h \
 		rtsp-client.h \
-		rtsp-server.h
+		rtsp-server-wfd.h \
+		rtsp-server.h \
+		gstwfdmessage.h
 
 c_sources = \
 	rtsp-auth.c \
@@ -28,6 +32,7 @@
 	rtsp-thread-pool.c \
 	rtsp-media.c \
 	rtsp-media-factory.c \
+	rtsp-media-factory-wfd.c \
 	rtsp-media-factory-uri.c \
 	rtsp-mount-points.c \
 	rtsp-permissions.c \
@@ -37,7 +42,10 @@
 	rtsp-session-media.c \
 	rtsp-session-pool.c \
 	rtsp-token.c \
+	gstwfdmessage.c \
+	rtsp-client-wfd.c \
 	rtsp-client.c \
+	rtsp-server-wfd.c \
 	rtsp-server.c
 
 noinst_HEADERS = 
diff -urN /home/zzoon.ko/opensrc/tarball/gst-rtsp-server-1.2.3/gst/rtsp-server/rtsp-client.c ./rtsp-client.c
--- /home/zzoon.ko/opensrc/tarball/gst-rtsp-server-1.2.3/gst/rtsp-server/rtsp-client.c	2014-02-10 21:43:30.000000000 +0900
+++ ./rtsp-client.c	2015-02-23 16:33:37.059801061 +0900
@@ -137,6 +137,12 @@
     GstRTSPContext * ctx);
 static gchar *default_make_path_from_uri (GstRTSPClient * client,
     const GstRTSPUrl * uri);
+static gboolean default_handle_options_request (GstRTSPClient * client,
+    GstRTSPContext * ctx);
+static gboolean default_handle_set_param_request (GstRTSPClient * client,
+    GstRTSPContext * ctx);
+static gboolean default_handle_get_param_request (GstRTSPClient * client,
+    GstRTSPContext * ctx);
 
 G_DEFINE_TYPE (GstRTSPClient, gst_rtsp_client, G_TYPE_OBJECT);
 
@@ -159,6 +165,9 @@
   klass->params_set = default_params_set;
   klass->params_get = default_params_get;
   klass->make_path_from_uri = default_make_path_from_uri;
+  klass->handle_options_request = default_handle_options_request;
+  klass->handle_set_param_request = default_handle_set_param_request;
+  klass->handle_get_param_request = default_handle_get_param_request;
 
   g_object_class_install_property (gobject_class, PROP_SESSION_POOL,
       g_param_spec_object ("session-pool", "Session Pool",
@@ -869,7 +878,7 @@
 }
 
 static gboolean
-handle_get_param_request (GstRTSPClient * client, GstRTSPContext * ctx)
+default_handle_get_param_request (GstRTSPClient * client, GstRTSPContext * ctx)
 {
   GstRTSPResult res;
   guint8 *data;
@@ -906,7 +915,7 @@
 }
 
 static gboolean
-handle_set_param_request (GstRTSPClient * client, GstRTSPContext * ctx)
+default_handle_set_param_request (GstRTSPClient * client, GstRTSPContext * ctx)
 {
   GstRTSPResult res;
   guint8 *data;
@@ -1462,6 +1471,8 @@
   if (media == NULL)
     goto media_not_found_no_reply;
 
+  /* FIXME-WFD : wfd url problem */
+#if 0
   if (path[matched] == '\0')
     goto control_not_found;
 
@@ -1469,6 +1480,9 @@
   path[matched] = '\0';
   /* control is remainder */
   control = &path[matched + 1];
+#else
+  control = g_strdup ("stream=0");
+#endif
 
   /* find the stream now using the control part */
   stream = gst_rtsp_media_find_stream (media, control);
@@ -1534,6 +1548,10 @@
   /* create and serialize the server transport */
   st = make_server_transport (client, ctx, ct);
   trans_str = gst_rtsp_transport_as_text (st);
+
+  /* FIXME-WFD : Temporarily force to set profile string */
+  trans_str = g_strjoinv ("RTP/AVP/UDP", g_strsplit (trans_str, "RTP/AVP", -1));
+
   gst_rtsp_transport_free (st);
 
   /* construct the response now */
@@ -1820,7 +1838,7 @@
 }
 
 static gboolean
-handle_options_request (GstRTSPClient * client, GstRTSPContext * ctx)
+default_handle_options_request (GstRTSPClient * client, GstRTSPContext * ctx)
 {
   GstRTSPMethod options;
   gchar *str;
@@ -1906,6 +1924,9 @@
   GstRTSPContext sctx = { NULL }, *ctx;
   GstRTSPMessage response = { 0 };
   gchar *sessid;
+  GstRTSPClientClass *klass;
+
+  klass = GST_RTSP_CLIENT_GET_CLASS (client);
 
   if (!(ctx = gst_rtsp_context_get_current ())) {
     ctx = &sctx;
@@ -1995,7 +2016,7 @@
   /* now see what is asked and dispatch to a dedicated handler */
   switch (method) {
     case GST_RTSP_OPTIONS:
-      handle_options_request (client, ctx);
+      klass->handle_options_request (client, ctx);
       break;
     case GST_RTSP_DESCRIBE:
       handle_describe_request (client, ctx);
@@ -2013,10 +2034,10 @@
       handle_teardown_request (client, ctx);
       break;
     case GST_RTSP_SET_PARAMETER:
-      handle_set_param_request (client, ctx);
+      klass->handle_set_param_request (client, ctx);
       break;
     case GST_RTSP_GET_PARAMETER:
-      handle_get_param_request (client, ctx);
+      klass->handle_get_param_request (client, ctx);
       break;
     case GST_RTSP_ANNOUNCE:
     case GST_RTSP_RECORD:
diff -urN /home/zzoon.ko/opensrc/tarball/gst-rtsp-server-1.2.3/gst/rtsp-server/rtsp-client.h ./rtsp-client.h
--- /home/zzoon.ko/opensrc/tarball/gst-rtsp-server-1.2.3/gst/rtsp-server/rtsp-client.h	2014-01-13 01:50:54.000000000 +0900
+++ ./rtsp-client.h	2015-02-17 15:57:28.047757756 +0900
@@ -102,6 +102,10 @@
   GstRTSPResult   (*params_get) (GstRTSPClient *client, GstRTSPContext *ctx);
   gchar *         (*make_path_from_uri) (GstRTSPClient *client, const GstRTSPUrl *uri);
 
+  gboolean        (*handle_options_request) (GstRTSPClient * client, GstRTSPContext * ctx);
+  gboolean        (*handle_set_param_request) (GstRTSPClient * client, GstRTSPContext * ctx);
+  gboolean        (*handle_get_param_request) (GstRTSPClient * client, GstRTSPContext * ctx);
+
   /* signals */
   void     (*closed)                  (GstRTSPClient *client);
   void     (*new_session)             (GstRTSPClient *client, GstRTSPSession *session);
@@ -177,7 +181,7 @@
                                                                 GstRTSPSession *sess,
                                                                 gpointer user_data);
 
-GList *                gst_rtsp_client_session_filter    (GstRTSPClient *client,
+GList *               gst_rtsp_client_session_filter    (GstRTSPClient *client,
                                                           GstRTSPClientSessionFilterFunc func,
                                                           gpointer user_data);
 
diff -urN /home/zzoon.ko/opensrc/tarball/gst-rtsp-server-1.2.3/gst/rtsp-server/rtsp-media.c ./rtsp-media.c
--- /home/zzoon.ko/opensrc/tarball/gst-rtsp-server-1.2.3/gst/rtsp-server/rtsp-media.c	2014-02-22 00:11:11.000000000 +0900
+++ ./rtsp-media.c	2015-02-17 15:57:25.483757666 +0900
@@ -241,10 +241,12 @@
           "If this media pipeline can be reused after an unprepare",
           DEFAULT_REUSABLE, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
 
+#ifdef GST_TYPE_RTSP_PROFILE
   g_object_class_install_property (gobject_class, PROP_PROFILES,
       g_param_spec_flags ("profiles", "Profiles",
           "Allowed transfer profiles", GST_TYPE_RTSP_PROFILE,
           DEFAULT_PROFILES, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+#endif
 
   g_object_class_install_property (gobject_class, PROP_PROTOCOLS,
       g_param_spec_flags ("protocols", "Protocols",
diff -urN /home/zzoon.ko/opensrc/tarball/gst-rtsp-server-1.2.3/gst/rtsp-server/rtsp-stream.c ./rtsp-stream.c
--- /home/zzoon.ko/opensrc/tarball/gst-rtsp-server-1.2.3/gst/rtsp-server/rtsp-stream.c	2014-02-22 00:11:11.000000000 +0900
+++ ./rtsp-stream.c	2015-02-17 15:57:25.527757667 +0900
@@ -173,10 +173,12 @@
           "The control string for this stream", DEFAULT_CONTROL,
           G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
 
+#ifdef GST_TYPE_RTSP_PROFILE
   g_object_class_install_property (gobject_class, PROP_PROFILES,
       g_param_spec_flags ("profiles", "Profiles",
           "Allowed transfer profiles", GST_TYPE_RTSP_PROFILE,
           DEFAULT_PROFILES, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+#endif
 
   g_object_class_install_property (gobject_class, PROP_PROTOCOLS,
       g_param_spec_flags ("protocols", "Protocols",
@@ -1001,6 +1003,9 @@
       inetaddr = g_inet_address_new_any (family);
   }
 
+  /* FIXME-WFD : Force to set 19000 as port number */
+  tmp_rtp = 19000;
+
   rtp_sockaddr = g_inet_socket_address_new (inetaddr, tmp_rtp);
   if (!g_socket_bind (rtp_socket, rtp_sockaddr, FALSE, NULL)) {
     g_object_unref (rtp_sockaddr);
@@ -1171,9 +1176,14 @@
       G_SOCKET_FAMILY_IPV4, priv->udpsrc_v4, priv->udpsink,
       &priv->server_port_v4, &priv->server_addr_v4);
 
+  /* FIXME-WFD : force to disable ipv6 mode in WFD mode */
+#if 0
   priv->have_ipv6 = alloc_ports_one_family (priv->pool, priv->buffer_size,
       G_SOCKET_FAMILY_IPV6, priv->udpsrc_v6, priv->udpsink,
       &priv->server_port_v6, &priv->server_addr_v6);
+#else
+  priv->have_ipv6 = FALSE;
+#endif
 
   return priv->have_ipv4 || priv->have_ipv6;
 }
@@ -1552,6 +1562,8 @@
 
   /* get the session */
   g_signal_emit_by_name (rtpbin, "get-internal-session", idx, &priv->session);
+  GST_DEBUG("zzoon unref internal session");
+  g_object_unref(priv->session);
 
   g_signal_connect (priv->session, "on-new-ssrc", (GCallback) on_new_ssrc,
       stream);
@@ -1841,6 +1853,7 @@
   gst_object_unref (priv->send_src[1]);
   priv->send_src[1] = NULL;
 
+  GST_DEBUG("Unref rtp session");
   g_object_unref (priv->session);
   priv->session = NULL;
   if (priv->caps)
